New quadplane landing approach system

tridge opened the pull request:
This implements a new quadplane landing approach system.
Key features:
 - adds a new QPOS_APPROACH stage for a NAV_VTOL_LAND waypoint, which is done as fixed wing. Similar to what RTL/QRTL does it will automatically switch to VTOL once the stopping distance calculated from Q_TRANS_DECEL is reached
 - adds a new QPOS_AIRBRAKE stage after QPOS_APPROACH which uses the VTOL motors to create drag to slow the aircraft in preparation for QPOS_POSITION1 stage (airbrake stage is skipped for tailsitters)
 - new velocity profile for QPOS_POSITION1 based on work done by @lukedepmsey in #14778 
 - new path to switching from QPOS_POSITION1 to QPOS_POSITION2 based on low velocity, to cope with the possibility that the fwd motor is not working
 - allow start of descent when distance remaining is below 2*height. This makes for a much smoother landing, and saves us spending a long time on the final piece of the positiioning (wasting battery)

In the RF9 AltiTransition the overall battery savings on landing is around 30%. It also means that users don't need to carefully setup the transition waypoint for a VTOL_LAND in AUTO. They can setup the mission to start the land way back and it will fly as fixed wing until it needs to transition.

The new stages apply both to VTOL_LAND and to QRTL, which means you can now do QRTL from a long way away and it will fly as fixed wing until the right transition point. Previously you needed to use Q_RTL_MODE=1 to achieve that.

UPDATE: This PR is now rebased on Leonards pending PR #17345

WickedShell commented on ArduPlane/quadplane.cpp:
```diff
@@ -2267,50 +2288,109 @@ void QuadPlane::vtol_position_controller(void)
         float target_speed = target_speed_xy.length();
+
+        // if vehicle is within deceleration profile, update target speed
+        if (poscontrol.stopping_distance > distance) {
```
Why do we stash this value instead of calculating it on demand? I don't see any reason we should be assuming the stopping distance calculation is still correct. If we were approaching from 1 km away the wind conditions could be very different and the distance calculation could be completely wrong, not to mention we may have been going the wrong way when we first calculated it and have had a tail wind, but by the time we reach here it we may have switched to a head wind which would cause us to detransition far to early (the other way around is even worse for transitioning)

tridge replied:
This was an adjustment made by Luke Dempsey and was for the previous code where we may have started POSITION1 before the stopping distance. I've removed it now, as we don't start POSITION1 until we've reached the stopping distance

tridge replied:
I think this still needs more work though, to account for the "going the wrong way" case

tridge replied:
fixed

WickedShell commented:
This is more an alternative of https://github.com/WickedShell/ardupilot/tree/wickedshell/vtol-land-straight the main thing this is missing that that allows is toggling of the type of landing on a per waypoint basis.

The changes to QRTL need to be opt in, because I know of a couple of people who are using the fact that the quad motors have to come on instantly and are relying upon the quad system rather then using the forward flight regime to get there. (It's mostly for failsafe reasons when the forward motor is out of power/gas and the VTOL system needs to be used)

tridge commented on ArduPlane/quadplane.cpp:
```diff
@@ -2267,50 +2288,109 @@ void QuadPlane::vtol_position_controller(void)
+        if (poscontrol.stopping_distance <= 0) {
+            // Calculate stopping distance based on the current groundspeed
+            // and the transition deceleration target
+            poscontrol.stopping_distance = groundspeed.length_squared()/(2*transition_decel);
```
fixed

tridge commented on ArduPlane/quadplane.cpp:
```diff
@@ -2267,50 +2288,109 @@ void QuadPlane::vtol_position_controller(void)
+        const float airbrake_time = 2;
+        if (!is_tailsitter() &&
+             poscontrol.state == QPOS_APPROACH &&
+            distance < stop_distance + airbrake_time*groundspeed) {
```
it falls thru to the normal stopping distance test as well, so should go straight to QPOS_POSITION1 if we start this distance <= stopping distance
I agree though that a helper would be good, and would allow us to use airbrake in the circle-to-land code and QRTL

tridge commented on ArduPlane/quadplane.cpp:
```diff
@@ -2859,14 +2940,19 @@ bool QuadPlane::verify_vtol_land(void)
     if (!available()) {
         return true;
     }
+    float height_above_ground = plane.relative_ground_altitude(plane.g.rangefinder_landing);
```
const added on all of these calls

tridge commented on ArduPlane/quadplane.cpp:
```diff
@@ -2859,14 +2940,19 @@ bool QuadPlane::verify_vtol_land(void)
     if (poscontrol.state == QPOS_POSITION2 &&
-        plane.auto_state.wp_distance < 2) {
+        (plane.auto_state.wp_distance < 2 ||
+         height_above_ground > descent_height_ratio*plane.auto_state.wp_distance)) {
```
this is a tricky one. I wanted to stop us waiting forever if a bit of wind means we're taking a long time getting to the descent point. Good one to discuss in more depth

tridge commented on ArduPlane/quadplane.cpp:
```diff
@@ -2267,50 +2288,109 @@ void QuadPlane::vtol_position_controller(void)
+        plane.calc_nav_roll();
+
+        // see if we have reached the transition point
+        uint16_t qrtl_radius = abs(plane.g.rtl_radius);
```
I've removed the use of rtl_radius here, it was a copy of the RTL -> QRTL logic which is needed there to stop it entering an orbit, but doesn't make sense here

tridge commented:
> The changes to QRTL need to be opt in, because I know of a couple of people who are using the fact that the quad motors have to come on instantly and are relying upon the quad system rather then using the forward flight regime to get there. (It's mostly for failsafe reasons when the forward motor is out of power/gas and the VTOL system needs to be used)

I've added in a check for not making progress in the approach to detect failed fwd motor. I'd like the approach and airbrake systems to be on by default if we can make the fwd thrust failure detection robust enough

pompecukor commented:
@tridge 

How will you hand failed transition. So QRTL to Fixed wing and we hit transition fail (for instance time limit), then what? Will it continue to attempt it infinity?

tridge commented:
> How will you handle failed transition? So QRTL to Fixed wing and say we hit "transition fail" (for instance time limit and FBWA not reached, etc), then what? Will it continue to attempt it infinity?

this PR is about landing, so the transitions are from fixed wing to vtol. If we're in vtol at the time and we fail fwd transition then I think the only choice would be QLAND. It is a good point to think about these cases.

Hwurzburg commented:
How does this impact Q_OPTION= bit4? "Use fixed wing approach"

tridge commented:
> How will you handle failed transition? So QRTL to Fixed wing and say we hit "transition fail" (for instance time limit and FBWA not reached, etc), then what? Will it continue to attempt it infinity?

it detects the lack of fwd progress and starts flying VTOL

Hwurzburg commented:
began SITL testing the QRTL changes...started off with Convergence...observations:
1. really like this behaviour!
2. really do not like the fixed wing portion of the return going below ALT_HOLD_RTL while a considerable distance from HOME which you can get if low and/or far away....would prefer it be constrained to above that until RTL_RADIUS as it being the safe height for a normal RTL...even though we talked about that being too high sometimes....for cases when you are below that alt and/or farther out and do QRTL....had several approaches at some distance beyond the normal RTL radius close to the much lower Q_RTL_ALT while still in fixed wing...or even a staged constraint....maybe 2-3X RTL_RADIUS, dont let it go below fixed wing ALT_HOLD_RTL...to provide some return height safety margin
another bad scenario.....flew out a ways and dropped down behind some trees or hillside in VTOL and did a QRTL FS....flew back at 15m.....
3. its pretty funky having it start fixed wing transition if you do a close in QRTL....moving away from HOME and then reverting..plus it does not rise to Q_RTL_ALT until it reverts if down low, so could crash into a close obstacle..perhaps a new radius param that will prevent that glitch if within it? or hard coded to 30 or 40 meters?should be able to VTOL RTL if already that close...
4. shouldnt this be enabled with the Q_RTL_MODE param rather than a bit in Q_OPTIONS?

-Tested with all Q_RTL_MODE options....QRTL portion works as if it was manually selected instead of automatically (ie as expected)
-Tested Alti with forward motor loss during fixed wing return, **_crashes_** into the ground, ie glides until crashes, never switches to VTOL...probably since forward progress continues in a glide....probably need to force an altitude QASSIST if it goes below target alt

pompecukor commented:
@tridge

Hi, we need a workable option for users that have spring/centered throttle with quad plane. Quite a popular and usefull option these days.

Thank you.
Alex

tridge commented:
>     2\. really do not like the fixed wing portion of the return going below ALT_HOLD_RTL while a considerable distance from HOME which you can get if low and/or far away.

i've implemented this now

>     3\. ts pretty funky having it start fixed wing transition if you do a close in QRTL....moving away from HOME and then reverting

I've fixed it so if within 1.5*radius and already in a VTOL state then it avoids the transition

tridge commented:
> Tested Alti with forward motor loss during fixed wing return, **_crashes_** into the ground, ie glides until crashes, never switches to VTOL...probably since forward progress continues in a glide....probably need to force an altitude QASSIST if it goes below target alt

I've fixed this with a new thrust loss detector

WickedShell commented on libraries/AP_Math/AP_Math.cpp:
```diff
@@ -82,13 +82,28 @@ template float safe_sqrt<short>(const short v);
                          float var_value,
                          float var_low, float var_high)
 {
+    if (var_low > var_high) {
```
Why are we getting into this situation? Given how much we use linear interpolate it seems a bit worrisome to

A: hit the situation
B: it seems like an extra cost we shouldn't be paying...

tridge replied:
we can use user settable parameters as the two endpoints we are interpolating between, and we want to ensure we get sane behaviour if the two are in an unexpected order. For example, its valid for a user to setup landing approach speed to be higher than cruise speed. 
The other key thing is API robustness, where the programmer needs to think carefully about increasing vs decreasing variable. 
even though I originally added linear_interpolate() I had forgotten it wasn't symmetric and made a mistake in the ordering. I don't want others to make the same mistake

WickedShell commented on ArduPlane/quadplane.cpp:
```diff
@@ -2557,7 +2586,128 @@ void QuadPlane::vtol_position_controller(void)
+        // properly spin up
+        const uint32_t min_airbrake_ms = 1000;
+        if (poscontrol.get_state() == QPOS_AIRBRAKE &&
+            poscontrol.time_since_state_start_ms() > min_airbrake_ms &&
```
We should probably do the time on this here rather then in the function, with millis being computed earlier as a const, as we are doing it in here + inside the body + outside in the next body. I guess the tradeoff is that requires not hiding the change time inside the private part of the struct....

tridge replied:
right, I wanted to hide that, I'm not really worried about the extra ms call (plus it prevents a race with a state change inside this function)

WickedShell reviewed (commented):
I haven't gotten to sim it at all yet, I'm still hoping to get to that today, but I wanted to post the initial comments I had (which isn't a full review, it's been a rather jumping around one), while I tried to find time to actually sim some approaches.

tridge commented on ArduPlane/quadplane.cpp:
```diff
@@ -2543,6 +2553,25 @@ void QuadPlane::run_xy_controller(void)
+ */
+void QuadPlane::poscontrol_init_approach(void)
+{
+    poscontrol.start_closing_vel = landing_closing_velocity().length();
```
i've removed this

tridge commented on ArduPlane/quadplane.cpp:
```diff
@@ -2543,6 +2553,25 @@ void QuadPlane::run_xy_controller(void)
+void QuadPlane::poscontrol_init_approach(void)
+{
+    poscontrol.start_closing_vel = landing_closing_velocity().length();
+    poscontrol.start_dist = plane.current_loc.get_distance(plane.next_WP_loc);
```
fixed

tridge commented on ArduPlane/quadplane.cpp:
```diff
@@ -2557,7 +2586,128 @@ void QuadPlane::vtol_position_controller(void)
+
+        bool throttle_saturated = SRV_Channels::get_output_scaled(SRV_Channel::k_throttle) >= plane.aparm.throttle_max;
+        if (throttle_saturated &&
+            motors->get_desired_spool_state() < AP_Motors::DesiredSpoolState::THROTTLE_UNLIMITED &&
```
added comment to clarify

tridge commented on ArduPlane/quadplane.cpp:
```diff
@@ -2557,7 +2586,128 @@ void QuadPlane::vtol_position_controller(void)
+
+        /*
+          we must switch to POSITION1 if our airspeed drops below the
+          assist speed. We additionally switch tp POSITION1 if we are
```
fixed

tridge commented on ArduPlane/quadplane.cpp:
```diff
@@ -2557,7 +2586,128 @@ void QuadPlane::vtol_position_controller(void)
+        const float closing_speed = closing_vel.length();
+        const float desired_closing_speed = desired_closing_vel.length();
+        if (!plane.ahrs.airspeed_estimate(aspeed)) {
+            aspeed = groundspeed;
```
this case won't happen in practice, as we have an airspeed estimate even with no airspeed, so this covers the case the user didn't wait for GPS lock and then somehow ended up in a transition

tridge commented on ArduPlane/quadplane.cpp:
```diff
@@ -2557,7 +2586,128 @@ void QuadPlane::vtol_position_controller(void)
+        const Vector2f desired_closing_vel = landing_desired_closing_velocity();
+        const float groundspeed = plane.ahrs.groundspeed();
+        const float distance = plane.auto_state.wp_distance;
+        const float closing_speed = closing_vel.length();
```
the only impact of closing_speed being off is that we may switch to POSITION1 early, which is benign. AIRBRAKE is an optimisation which we can afford to lose if the setup for the approach went badly wrong
